### ISSUE TITLE
pam_modutil_getpwnam() leak

### ISSUE THREAD

**Issue by AlessandroBono** [linux-pam/linux-pam#482]
While it is unlikely that `pam_modutil_getpwnam()` gets called with `pamh == NULL`, it looks like that when we hit this case [here](https://github.com/linux-pam/linux-pam/blob/367bcfa31307379b9eeabf63af31b48e63ffdbb5/libpam/pam_modutil_getpwnam.c#L86) the returned structure is leaked.

Is this an expected behavior?

**Comment by ldv-alt** [comment]
The structure is being returned by `pam_modutil_getpwnam()` to the caller, isn't it?
Anyway, since the function arguments are declared as `PAM_NONNULL((1,2))`, I wonder why the function doesn't just return NULL in case of `pamh == NULL`.

**Comment by AlessandroBono** [comment]
> The structure is being returned by pam_modutil_getpwnam() to the caller, isn't it?

Yes, but in that case the caller has to free the structure, while in the other case we use `pam_modutil_cleanup` to do that.

**Comment by t8m** [comment]
We could document the need to deallocate the structure if called with NULL pamh and allow that. However there is a problem with the `ifndef HAVE_GETPWNAM_R` case.

**Comment by cgzones** [comment]
Since `pam_modutil_getpwnam()` is annotated with `PAM_NONNULL((1,2))` calling it with `pamh` of NULL is undefined behavior, and compilers are allowed to drop all `pamh != NULL` checks inside of `pam_modutil_getpwnam()` away, so the mentioned branch might well be optimized away.

**Comment by ldv-alt** [comment]
`PAM_NONNULL` is defined to nothing inside libpam.
